### PR TITLE
(PUP-8121) Make #new function signatures more strict

### PR DESCRIPTION
**Pull request by thallgren**
The `Init` data type bases assignability and instance computation on the
signatures of the #new function for a given type. Before this commit,
those signatures were very relaxed. An `Integer` could for instance be
created from any `String` which gave that "blue" was an instance of
`Init[Integer]`. This commit tightens the constraints.

### CODE REVIEW

File: lib/puppet/pops/types/types.rb [hunk 1]
```diff
@@ -1108,15 +1127,28 @@ def from_convertible(from, radix)
...
+      def on_error(from, radix = :default, abs = nil)
+        assert_radix(radix) unless radix == :default
+        t = TypeCalculator.singleton.infer(from).generalize
+        "Value of type '#{t}' cannot be converted to Integer"
+      end
+
       def assert_radix(radix)
         case radix
-        when 2, 8, 10, 16, :default
+        when 2, 8, 10, 16
```
Comment (hlindberg) [inline lib/puppet/pops/types/types.rb]: 'default' removed while error message below says it is acceptable. Is there any harm in keeping the literal "default" here for the radix parameter?

File: lib/puppet/functions/new.rb [hunk 2]
```diff
@@ -14,10 +14,7 @@
 
   def new_instance(scope, t, *args)
     return args[0] if args.size == 1 && !t.is_a?(Puppet::Pops::Types::PInitType) && t.instance?(args[0])
-    result = catch :undefined_value do
-      new_function_for_type(t, scope).call(scope, *args)
-    end
-    assert_type(t, result)
+    result = assert_type(t, new_function_for_type(t, scope).call(scope, *args))
```
Comment (hlindberg) [inline lib/puppet/functions/new.rb]: the catching of :undefined_value is removed - there is still logic that throws that and I think the result becomes  a bad error (unhandled throw of :undefied_value) instead of the error produced by assert_type.

File: lib/puppet/pops/types/p_sem_ver_type.rb [hunk 3]
```diff
@@ -97,6 +101,10 @@ def from_args(major, minor, patch, prerelease = nil, build = nil)
       def from_hash(hash)
         SemanticPuppet::Version.new(hash['major'], hash['minor'], hash['patch'], hash['prerelease'], hash['build'])
       end
+
+      def on_error(str)
+        "The string '#{str}' cannot be converted to a SemVer"
```
Comment (hlindberg) [inline lib/puppet/pops/types/p_sem_ver_type.rb]: None of the errors are handled with I18N support. (The earlier version did not have I18N either, but it should as these type conversion errors are user facing).  Will mark other occurrences with only "i18n"

File: lib/puppet/pops/types/types.rb [hunk 4]
```diff
@@ -1108,15 +1127,28 @@ def from_convertible(from, radix)
...
+        end
+        radix = args_hash['radix']
+        assert_radix(radix) unless radix.nil? || radix == :default
+        TypeAsserter.assert_instance_of('Integer.new', loader.load(:type, 'namedargs'), args_hash)
+      end
+
+      def on_error(from, radix = :default, abs = nil)
+        assert_radix(radix) unless radix == :default
+        t = TypeCalculator.singleton.infer(from).generalize
+        "Value of type '#{t}' cannot be converted to Integer"
```
Comment (hlindberg) [inline lib/puppet/pops/types/types.rb]: i18n

File: lib/puppet/pops/types/types.rb [hunk 5]
```diff
@@ -1223,11 +1258,13 @@ def from_convertible(from)
...
           end
-        else
-          t = Puppet::Pops::Types::TypeCalculator.singleton.infer(from).generalize
-          raise TypeConversionError.new("Value of type '#{t}' cannot be converted to Float")
         end
       end
+
+      def on_error(from, _)
+        t = TypeCalculator.singleton.infer(from).generalize
+        "Value of type '#{t}' cannot be converted to Float"
```
Comment (hlindberg) [inline lib/puppet/pops/types/types.rb]: i18n

File: lib/puppet/pops/types/types.rb [hunk 6]
```diff
@@ -1816,28 +1853,31 @@ def instance?(o, guard = nil)
...
-          true
         else
-          raise TypeConversionError.new("Value '#{from}' of type '#{from.class}' cannot be converted to Boolean")
+          true
         end
       end
+
+      def on_error(from)
+        t = TypeCalculator.singleton.infer(from).generalize
+        "Value of type '#{t}' cannot be converted to Boolean"
```
Comment (hlindberg) [inline lib/puppet/pops/types/types.rb]: i18n

File: lib/puppet/pops/types/types.rb [hunk 7]
```diff
@@ -2529,41 +2569,43 @@ def iterable_type(guard = nil)
...
-              raise TypeConversionError.new("Value of type '#{t}' cannot be converted to Array")
-            end
-          end
+          Iterable.on(from).to_a
         end
       end
+
+      def on_error(from, _ = false)
+        t = TypeCalculator.singleton.infer(from).generalize
+        "Value of type '#{t}' cannot be converted to Array"
```
Comment (hlindberg) [inline lib/puppet/pops/types/types.rb]: i18n

File: spec/unit/functions/new_spec.rb [hunk 8]
```diff
@@ -545,7 +545,7 @@
       expect{compile_to_catalog(<<-"MANIFEST"
         $x = Boolean.new(undef)
       MANIFEST
-      )}.to raise_error(Puppet::Error, /expects a Boolean value, got Undef/)
+      )}.to raise_error(Puppet::Error, /cannot be converted to Boolean/)
```
Comment (hlindberg) [inline spec/unit/functions/new_spec.rb]: should check that `undef` is mentioned in the error

File: spec/unit/functions/new_spec.rb [hunk 9]
```diff
@@ -28,7 +28,7 @@
       $x = Integer.new(undef)
       notify { "one${x}word": }
     MANIFEST
-    )}.to raise_error(Puppet::Error, /expects an Integer value, got Undef/)
+    )}.to raise_error(Puppet::Error, /cannot be converted to Integer/)
```
Comment (hlindberg) [inline spec/unit/functions/new_spec.rb]: should check that undef is mentioned in the error

Comment (hlindberg) [reply spec/unit/functions/new_spec.rb]: this as opposed to getting an empty string in the output or no clue at all what was wrong.

File: spec/unit/pops/types/p_init_type_spec.rb [hunk 10]
```diff
@@ -279,6 +279,64 @@ module Types
       CODE
       expect(eval_and_collect_notices(code)).to eql(["One({'init_one' => 'w'})"])
     end
+
+    context 'computes if x is an instance such that' do
+      %w(true false True False TRUE FALSE Yes No yes no YES NO y n Y N).each do |str|
```
Comment (hlindberg) [inline spec/unit/pops/types/p_init_type_spec.rb]: could add mixed case `fAlSe` ?

File: spec/unit/pops/types/p_init_type_spec.rb [hunk 11]
```diff
@@ -279,6 +279,64 @@ module Types
...
+        it "string '#{str}' is an instance of Init[Boolean]" do
+          expect(eval_and_collect_notices("notice('#{str}' =~ Init[Boolean])")).to eql(['true'])
+        end
+      end
+
+      it 'arbitrary string is not an instance of Init[Boolean]' do
+        expect(eval_and_collect_notices("notice('blue' =~ Init[Boolean])")).to eql(['false'])
+      end
+
```
Comment (hlindberg) [inline spec/unit/pops/types/p_init_type_spec.rb]: testing that empty string is not convertible to boolean false is of interest.

File: spec/unit/pops/types/p_init_type_spec.rb [hunk 12]
```diff
@@ -279,6 +279,64 @@ module Types
...
+      %w(1.2.3 1.1.1-a3 1.2.3+b3 1.2.3-a3+b3).each do |str|
+        it "string '#{str}' is an instance of Init[SemVer]" do
+          expect(eval_and_collect_notices("notice('#{str}' =~ Init[SemVer])")).to eql(['true'])
+        end
+      end
+
+      it 'non SemVer compliant string is not an instance of Init[SemVer]' do
+        expect(eval_and_collect_notices("notice('blue' =~ Init[SemVer])")).to eql(['false'])
+      end
+    end
```
Comment (hlindberg) [inline spec/unit/pops/types/p_init_type_spec.rb]: without reviewing the rest of the init[t] unit test logic - are there tests for edge cases like empty string `''` and `undef` for all?

File: spec/unit/functions/new_spec.rb (hunk 9 above)
Comment (thallgren) [reply spec/unit/functions/new_spec.rb]: The full message says "A Value of type Undef cannot be converted to Integer".

File: lib/puppet/functions/new.rb (hunk 2 above)
Comment (thallgren) [reply lib/puppet/functions/new.rb]: Where is the logic that still throws that? I thought I had removed all of it.

File: lib/puppet/pops/types/types.rb (hunk 1 above)
Comment (thallgren) [reply lib/puppet/pops/types/types.rb]: No, 'default' is not removed. This method is however never called upon when the value is 'default'
